### DISCUSSION TITLE
Possibility to edit checkout fields and shipping options

### DISCUSSION THREAD

**Discussion by AlexandruN** [woocommerce/woocommerce#44933] [WooCommerce, Blocks, and Site Editing]
Hello,

I just started playing around with the blocks checkout to try and change our plugin to also support this and we hit some shortcomings and wanted to know if there is a roadmap for these:
- the possibility to add different fields to the checkout page (this would be easy to implement using localization, but the problem is that when displaying the fields, the iteration is done on the defaultFields so the only localization that works is for those fields, like rearranging or changing labels). I don't know if this is intentional or a bug.
- the possibility to change a field's type and add autocomplete on it. The fields are rendering either as text or Country/State. I would like the possibility to render them as select, ideally with a server request for the data with search/autocomplete.
- the possibility to add something after each shipping method (I saw that there is the possibility to add a block after all the shipping methods). For example we have fixed point delivery, and for a shipping method there could be 100s of locations, and currently if a method is selected we display a select under if for selecting the location, which is not possible with the new blocks.

Please let me know if I am mistaken and any of the above is possible and if not, if there is a roadmap for when this will be implemented.

I very much like the direction this is going!

Best regards,
Alex

**Comment by passatgt** [comment]
Hi! About your third point: you can turn on Local Pickup in settings, so your checkout page will look like this:

<img width="840" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/43415/191250ec-91f2-4e03-98df-7e96b3012034">

And then if you select Local Pickup, it will display shipping methods that support local pickup(add "local-pickup" to your shipping method class supports array). 

And they you can display a custom UI just below the local pickup methods:

<img width="817" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/43415/b8a222f1-47d0-4526-9f85-6e539c52c9a7">

This is done by adding the block into the '_woocommerce/checkout-pickup-options-block_' parent. At the moment it's not possible to add it just below a specific shipping method, however there is a description part, maybe later we could hook into this. I show my custom UI to select a pickup point this way. After the point is selected, you can update your shipping rate's meta data with these options:

`$rate->add_meta_data('pickup_address', 'Selceted pickup poit name & address');`

And this will show up on the sidebar order summary:

<img width="441" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/43415/ad7a4082-4997-4f9a-b5e9-e1163c6897f3">

**Comment by opr** [comment]
Hi @AlexandruN and thanks for opening this discussion and for your feedback.

> the possibility to add different fields to the checkout page (this would be easy to implement using localization, but the problem is that when displaying the fields, the iteration is done on the defaultFields so the only localization that works is for those fields, like rearranging or changing labels). I don't know if this is intentional or a bug.

Yes we are working on this, please see [our discussion about it](https://github.com/woocommerce/woocommerce-blocks/discussions/11173)

> the possibility to change a field's type and add autocomplete on it. The fields are rendering either as text or Country/State. I would like the possibility to render them as select, ideally with a server request for the data with search/autocomplete.

We don't have plans to make this change to existing fields. Could you share a use case for this?

> the possibility to add something after each shipping method (I saw that there is the possibility to add a block after all the shipping methods). For example we have fixed point delivery, and for a shipping method there could be 100s of locations, and currently if a method is selected we display a select under if for selecting the location, which is not possible with the new blocks.

Nothing is specifically planned yet, but I have noticed an increased demand for the ability to render in more areas than we currently allow. I plan to open a discussion about this, and I will keep your use case in mind.

If you want to limit your delivery options to specific points then that's a good use case for Local Pickup, which @passatgt mentioned above. Thank you @passatgt!

**Comment by AlexandruN** [comment]
Thanks a lot for both your answers @passatgt  and @opr .

Regarding the fields, I read the discussion (sorry, did not see it before writing). It is a good starting point and it would cover most of the cases. 

For editing the current fields the main use case is to provide search/autocomplete also for city, so when a county is selected, the city field would allow search for the localities in the given city. This is something that we do in the non blocks checkout. We also change address_1 label to street, and add search for it too (based on the city and county). It can also be done by hiding the fields and showing custom ones and taking care of syncing the data, but it would be something nice to have :).

Also if the behavior of a filed like postal code could be changed, we would be able to implement a feature that autocompletes the city/county and street based on the postal code (something that we do on the non blocks checkout). It could still be done with the blocks (adding some javascript on top of it), but would not be the most elegant way. Also it would have to be some javascript added on top of it for the search of city/street etc, by not having the possibility to alter the behavior of fields.

Anyway, creating the ability to have custom fields would allow developing all of the above in a way or another. I would extend it by not only having the ability to create custom fields based on a predetermined list, but also have the ability to insert custom components. That would be really powerful (like is done in the core for CountrySelect and StateSelect) and would simplify a lot custom use cases like above.

Regarding the 3rd point (with the display of custom data under shipping methods) it would be possible to do it like stated above by @passatgt  (thanks), but the main problem is that we could have 6-7 such services, and if the user selects the first one, it would be quite hard to indicate the need to select something after a list of other 6-7 items. Also, for each of the services, we might want to add a button to open a map and select it on the map. Showing the map button only after the methods would not be the easiest way to understand by a user I think, and would have to jump between the selections and the button, and based on this the button will show other locations on map. Also if you navigate with the keyboard (pressing tab), it would be quite frustrating. There are ways to work around this (show only a method, and when that is selected render our own list), but there would also need to be a customization for the price, to hide the price or show a custom text instead of it, since it will vary depending on the location selected in our rendered list.

**Reply by passatgt** [reply]
In my case, i have a single shipping method called Parcel Lockers(or you can name it anything you like of course) and it will show you options from multiple providers. So by default it looks like this:

<img width="801" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/43415/e501760f-bec3-4321-aef5-f9c3b2e0da5b">

You can see two provider icons(GLS and Foxpost). And if you click select pickup point, you'll see a map with both GLS and Foxpost options with different prices:

<img width="1237" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/43415/96f69e02-815b-4d05-b004-53d913289781">

And once a point is selected, then i show this UI:

<img width="661" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/43415/ff787c2a-e1ec-48d7-a8e1-52663645882d">

In your case if you have multiple shipping methods under Pickup Options, maybe you could add an additional step to the form, where you can select the specific pickup point. 

<img width="959" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/43415/6e40e5f1-6101-4774-8607-3c3536ef1e34">

**Reply by AlexandruN** [reply]
Thanks a lot @passatgt !

I will explore these options and they should be enough for what I need for the moment to circumvent the missing possibility of adding something right after the Shipping Method.
